cmd/testpeers.go

```diff
@@ -63,7 +108,145 @@ func supportedSelfTestCases() map[testCaseName]func(context.Context, *testPeersC
 	}
 }
 
-func runTestPeers(ctx context.Context, w io.Writer, cfg testPeersConfig) (err error) {
+func startTCPNode(ctx context.Context, cfg testPeersConfig) (host.Host, func(), error) {
+	var peers []p2p.Peer
+	for i, e := range cfg.ENRs {
+		record, err := enr.Parse(e)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "decode enr", z.Str("enr", e))
+		}
+
+		p, err := p2p.NewPeerFromENR(record, i)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		peers = append(peers, p)
+	}
+
+	key, err := p2p.LoadPrivKey(cfg.DataDir)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	r, err := enr.New(key)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	mePeer, err := p2p.NewPeerFromENR(r, len(cfg.ENRs))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	log.Info(ctx, "Self p2p name resolved", z.Any("name", mePeer.Name))
+
+	peers = append(peers, mePeer)
+
+	allENRs := cfg.ENRs
+	allENRs = append(allENRs, r.String())
+	slices.Sort(allENRs)
+	allENRsString := strings.Join(allENRs, ",")
+	allENRsHash := sha256.Sum256([]byte(allENRsString))
+
+	return setupP2P(ctx, key, cfg.P2P, peers, allENRsHash[:])
+}
+
+func setupP2P(ctx context.Context, key *k1.PrivateKey, cfg p2p.Config, peers []p2p.Peer, enrsHash []byte) (host.Host, func(), error) {
+	var peerIDs []peer.ID
+	for _, p := range peers {
+		peerIDs = append(peerIDs, p.ID)
+	}
+
+	if err := p2p.VerifyP2PKey(peers, key); err != nil {
+		return nil, nil, err
+	}
+
+	relays, err := p2p.NewRelays(ctx, cfg.Relays, hex.EncodeToString(enrsHash))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	connGater, err := p2p.NewConnGater(peerIDs, relays)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tcpNode, err := p2p.NewTCPNode(ctx, cfg, key, connGater, false)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	p2p.RegisterConnectionLogger(ctx, tcpNode, peerIDs)
+
+	for _, relay := range relays {
+		relay := relay
+		go p2p.NewRelayReserver(tcpNode, relay)(ctx)
+	}
+
+	go p2p.NewRelayRouter(tcpNode, peerIDs, relays)(ctx)
+
+	// Register peerinfo server handler for identification to relays (but do not run peerinfo client).
+	gitHash, _ := version.GitCommit()
+	_ = peerinfo.New(tcpNode, peerIDs, version.Version, enrsHash, gitHash, nil, false)
+
+	return tcpNode, func() {
+		err := tcpNode.Close()
+		if err != nil && !errors.Is(err, context.Canceled) {
+			log.Error(ctx, "Close TCP node", err)
+		}
+	}, nil
+}
+
+func pingPeerOnce(ctx context.Context, tcpNode host.Host, peer p2p.Peer) (ping.Result, error) {
```

Comment:
Wondering if we could reuse `PingService` here?

Reply:
If you refer to `NewPingService` in p2p package, it starts a long-lived pinging. What we can try to reuse is the `pingPeerOnce` function from the p2p package. We gotta export it alongside with some other functions in that package though.

Reply:
I'm not opposed to that. We could refactor `pingPeerOnce` as a standalone exported function and use it in both `PingService` and here. Not strictly necessary for this iteration, but would be a nice to have so that we use the same implementation in both.